lib/govuk_tech_docs/api_reference/api_reference_renderer.rb

```diff
@@ -167,108 +123,52 @@ def json_prettyprint(data)
       end
 
       def schema_properties(schema_data)
-        properties = Hash.new
-        if defined? schema_data.properties
-          schema_data.properties.each do |key, property|
-            properties[key] = property
-          end
-        end
-        properties.merge! get_all_of_hash(schema_data)
-        properties_hash = Hash.new
-        properties.each do |pkey, property|
-          if property.type == 'object'
-            properties_hash[pkey] = Hash.new
-            items = property.items
-            if !items.blank?
-              properties_hash[pkey] = schema_properties(items)
-            end
-            if !property.properties.blank?
-              properties_hash[pkey] = schema_properties(property)
-            end
-          elsif property.type == 'array'
-            properties_hash[pkey] = Array.new
-            items = property.items
-            if !items.blank?
-              properties_hash[pkey].push schema_properties(items)
-            end
-          else
-            properties_hash[pkey] = !property.example.nil? ? property.example : property.type
-          end
-        end
+        properties = schema_data.properties.to_h
 
-        properties_hash
-      end
-
-    private
-
-      def get_all_of_array(schema)
-        properties = Array.new
-        if schema.is_a?(Array)
-          schema = schema[1]
-        end
-        if schema["allOf"]
-          all_of = schema["allOf"]
+        schema_data.all_of.to_a.each do |all_of_schema|
+          properties.merge!(all_of_schema.properties.to_h)
         end
-        if !all_of.blank?
-          all_of.each do |schema_nested|
-            schema_nested.properties.each do |property|
-              if property.is_a?(Array)
-                property = property[1]
-              end
-              properties.push property
-            end
-          end
-        end
-        properties
-      end
 
-      def get_all_of_hash(schema)
-        properties = Hash.new
-        if schema["allOf"]
-          all_of = schema["allOf"]
+        properties.each_with_object({}) do |(name, schema), memo|
+          memo[name] = if schema.type == 'object'
+                         if schema.properties
+                           schema_properties(schema)
+                         elsif schema.items
+                           schema_properties(schema.items)
+                         else
+                           {}
+                         end
```

Comment:
This block feels like it could be refactored - something like:

```ruby
memo[name] = schema_properties(schema.items || schema)
memo[name] = (schema.example || schema.type) unless memo[name]
```

But I might be missing a nuance there - and appreciate that you're concerned about introducing new bugs.

Reply:
Thanks - I've not been very happy with my changes there. I don't dare quite go to the level you suggest here as there are a few more nuances. I've done a bit of refactoring to try make it all a bit easier to read though.